### PR TITLE
[RDY] Do not translate blank string

### LINKED ISSUE #16129: Random world name will always be "Tom" in Chinese version
Tested with zh-cn/zh-tw/ja, press \* will generate random name, and it will only be "tom" (and related translated word).

Random character name will also generate "Tom" a lot ("Tom" only, without last name),  some where around 1 in 5.

English version does not have this issue.

0.c-12763-g100d5bd ,windows

### LINKED ISSUE #17585: russian translation issue
pipe combination gun (комбинированный самопал) description contains strange text.

### PR DESCRIPTION
**Pull request by ghost**
This pull request fixes #17585. The `_()` function checks whether the given string is blank before actually calling `gettext()` function.

### CODE REVIEW

**Comment by codemime** [comment]
Please consider also extending the consistency check in `void Item_factory::check_definitions()`. Something like this:

```c++
if( type->description.empty() ) {
    msg << "empty description" << "\n";
}
```
and fill missing descriptions (according to the debug messages you get) to close the issue properly.

**Comment by DLaboratory** [comment]
@cainiaowu There are tens of thousands of world names in `en.json`. I can't endure translating those names. We must figure out a better way to make up Chinese world names fast and easy.

**Comment by illi-kun** [comment]
This PR contains of two unrelated fixes so it would be better to commit it separately (and split it in two PRs).

**Comment by DLaboratory** [comment]
OK, now this PR only contains the fix for #17585.

File: src/translations.h [hunk 1]
```diff
@@ -16,7 +16,13 @@
 #include <libintl.h>
 #include <clocale>
 
-#define _(STRING) gettext(STRING)
+extern char *blank_string;
+
+inline char *_( const char *msg )
+{
+    return ( msg[0] == '\0' ) ? blank_string : gettext( msg );
```
Comment (BevapDin) [inline src/translations.h]: Why so complicated? Just return a `const char *` and return `""` or the input string itself.

The other version that is used without `LOCALIZE` returns a `const char *` anyway.

Comment (DLaboratory) [inline src/translations.h]: Casting `const char*` to `char*` is prohibited.

Comment (BevapDin) [inline src/translations.h]: > Casting  const char*  to  char*  is prohibited.

Yes, that's why the function should return a `const char *`. The dummy macro that is used when `LOCALIZE` is not defined (see a few lines below) just forwards the input, which is already a `const char *` (it "returns" `const char *`) and quote from https://www.gnu.org/software/libc/manual/html_node/Translation-with-gettext.html

> Please note that although the return value is char * the returned string must not be changed. This broken type results from the history of the function and does not reflect the way the function should be used.

The result of `gettext` is effectively a `const char *`.

File: src/item_factory.cpp [hunk 2]
```diff
@@ -663,6 +663,9 @@ void Item_factory::check_definitions() const
         if( type->price < 0 ) {
             msg << "negative price" << "\n";
         }
+        if( type->description.size() == 0 ) {
```
Comment (BevapDin) [inline src/item_factory.cpp]: Why not use `std::string::empty()` instead of the explicit comparison?

**Pull request merged**